doc reference: add reference for "pgroonga_condition()" function (#368)
pgroonga/pgroonga.github.io: reference/functions/pgroonga-condition.md

                                    index_name => 'pgroonga_memos_index');
                                           QUERY PLAN                                            
-------------------------------------------------------------------------------------------------
 Seq Scan on memos  (cost=0.00..2.52 rows=1 width=100) (actual time=2.230..2.406 rows=2 loops=1)
   Filter: (title &@~ '(バイオリン,,,,pgroonga_memos_index,)'::pgroonga_condition)
   Rows Removed by Filter: 1
 Planning Time: 2.222 ms
 Execution Time: 2.525 ms
(5 rows)

SELECT *
  FROM memos
 WHERE title &@~ pgroonga_condition('バイオリン',
                                    index_name => 'pgroonga_memos_index');
 id |    title     |      content       | tag  
----+--------------+--------------------+------
  2 | ヴァイオリン | content2           | tag2
  1 | ヴァイオリン | Let's play violin! | 
(2 rows)
```

このようにして、シーケンシャルサーチ実行時でも、インデックスサーチ実行時でも検索結果が変わらないようにできます。

また、「タイトルは本文よりも重要」も実現できます。

例に使うサンプルスキーマとデータは次の通りです。

```sql
DROP TABLE IF EXISTS memos;
CREATE TABLE memos (
  title text,
  content text
);

CREATE INDEX pgroonga_memos_index
    ON memos
 USING pgroonga ((ARRAY[title, content]));
```

> ```

Reviewer: ```suggestion
```sql
```

他にも何個か追加したほうが良さそうなところがありました。

Author: おぉ。ありがとうございます！確認して追記します。

Author: 追加しました。

INSERT INTO memos VALUES ('PostgreSQL', 'PostgreSQLはリレーショナル・データベース管理システムです。');
INSERT INTO memos VALUES ('Groonga', 'Groongaは日本語対応の高速な全文検索エンジンです。');
INSERT INTO memos VALUES ('PGroonga', 'PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。');
INSERT INTO memos VALUES ('コマンドライン', 'groongaコマンドがあります。');
```

より指定したクエリーにマッチしたレコードを探すためには[pgroonga_score function][pgroonga-score-function]を使えます。

```
SELECT *, pgroonga_score(tableoid, ctid) AS score
  FROM memos
 WHERE ARRAY[title, content] &@~
       pgroonga_condition('Groonga OR PostgreSQL', ARRAY[5, 1])
 ORDER BY score DESC;
      title      |                                  content                                  | score 
----------------+---------------------------------------------------------------------------+-------
 Groonga        | Groongaは日本語対応の高速な全文検索エンジンです。                         |     6
 PostgreSQL     | PostgreSQLはリレーショナル・データベース管理システムです。                |     6
 PGroonga       | PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。 |     2
 コマンドライン | groongaコマンドがあります。                                               |     1
(4 rows)
```

> titleカラムに「Groonga」または「PostgreSQL」があるレコードの方がcontentカラムに「Groonga」または「PostgreSQL」がある方がスコアーが高いことを確認できます。

Reviewer: `また、「タイトルは本文よりも重要」も実現できます。` からの説明で、 `weights` についての説明をしていると思うので、 `weights` という用語を出しつつ説明したほうが良さげです。

「スコア（score）」という単語が多数登場するので、pgroonga_conditionのscorersの説明に見えました。

Author: なるべく`weight`という用語を出してみましたが、どうでしょうか？

## Usage

Here are sample schema and data:

```sql
CREATE TABLE memos (
  id integer,
  title text,
  content text
);

CREATE INDEX pgroonga_memos_index
          ON memos
       USING pgroonga (title)
        WITH (normalizers='NormalizerNFKC150("unify_katakana_v_sounds", true)');

INSERT INTO memos VALUES (1, 'ヴァイオリン', E'Let\'s play violin!');
```

インデックスサーチ時はインデックスに指定したオプションで検索条件をカスタマイズできます。

> 上のサンプルでは、`normalizers='...'`の部分でオプションを指定し、検索条件をカスタマイズしています。

Reviewer 2: ノーマライザーのオプションで変更することは「検索条件をカスタマイズ」とは言わないんじゃないかな。。。`AND x > 10`を増やすとかそういうときに使うんじゃないかな。

検索方法をカスタマイズかなぁ。ちょっと違う気もするけど。。。

Author: 次のように検索方法や検索条件という言葉を使わずに、ノーマライズ方法、トークナイズ方法をカスタマイズするという表現にするのはどうでしょうか？

```diff
-インデックスサーチ時はインデックスに指定したオプションで検索条件をカスタマイズできます。
-上のサンプルでは、`normalizers='...'`の部分でオプションを指定し、検索条件をカスタマイズしています。
+インデックスサーチ時はインデックスに指定したオプションを使ってノーマライズ方法やトークナイズ方法をカスタマイズできます。
+上のサンプルでは、`normalizers='...'`の部分でオプションを指定し、ノーマライズ方法をカスタマイズしています。
```

ノーマライズ、トークナイズという言葉を理解している人が読む前提になりますが、ここをカスタマイズする人はノーマライズやトークナイズについて知っていると思うので、上の表現でも伝わると考えました。

Reviewer 2: そんな感じにするならこうじゃない？

インデックス作成時にオプションを指定してインデックスサーチ時の挙動をカスタマイズできます。
上のサンプルでは、`normalizers='...'`の部分でオプションを指定しています。

Author: なるほど。↑の表現に変更しました。

一方、インデックスサーチではなくシーケンシャルサーチが実行されると、インデックスに指定されているオプションは参照できません。
シーケンシャルサーチ時はどのインデックスを参照すればよいかという情報がないからです。

インデックスを参照できないということは、インデックスに設定されているノーマライザーやトークナイザーの情報を参照できず、適用もできません。
そのためシーケンシャルサーチ実行時は、インデックスサーチ実行時と検索結果が異なる可能性があります。
この問題を回避するためにシーケンシャルサーチ時に参照するインデックスを明示的に指定します。
`pgroonga_condition()`の`index_name => '...'`がそのための引数です。

> 次の例は、シーケンシャルサーチになっていますが、「バイオリン」で「ヴァイオリン」がヒットしています。

Author: `remove_symbol`なら、英語でも通用する例ができそうなので、`remove_symbol`を使った例にしてみます。

> 次の例は、シーケンシャルサーチになっていますが、「バイオリン」で「ヴァイオリン」がヒットしています。

Author: `remove_symbol`を使った例に変更しました。

INSERT INTO memos VALUES ('PGroonga', 'PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。');
INSERT INTO memos VALUES ('コマンドライン', 'groongaコマンドがあります。');
```

より指定したクエリーにマッチしたレコードを探すためには[pgroonga_score function][pgroonga-score-function]を使えます。

```sql
SELECT *, pgroonga_score(tableoid, ctid) AS score
  FROM memos
 WHERE ARRAY[title, content] &@~
       pgroonga_condition('Groonga OR PostgreSQL', ARRAY[5, 1])
 ORDER BY score DESC;
      title      |                                  content                                  | score 
----------------+---------------------------------------------------------------------------+-------
 Groonga        | Groongaは日本語対応の高速な全文検索エンジンです。                         |     6
 PostgreSQL     | PostgreSQLはリレーショナル・データベース管理システムです。                |     6
 PGroonga       | PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。 |     2
 コマンドライン | groongaコマンドがあります。                                               |     1
(4 rows)
```

上の例では、`ARRAY[title, content] &@~ pgroonga_condition('Groonga OR PostgreSQL', ARRAY[5, 1])`と指定しているので、タイトルが本文より5倍重要としています。
titleカラムに「Groonga」または「PostgreSQL」があるレコードの方がcontentカラムに「Groonga」または「PostgreSQL」がある方よりスコアーが高いことを確認できます。

> カラム毎の`weight`を設定しつつ、シーケンシャルサーチ時でもインデックスサーチ時でも検索結果を同じにするためには、`pgroonga_condition('keyword', ARRAY[weight1, weight2, ...], index_name => 'pgroonga_index')`を使います。

Reviewer 2: シーケンシャルサーチ時は`weight`は使われないよね？
↑の例みたいに`score`でソートすると違う結果にならない？

シーケンシャルサーチだとスコアーは再現できないみたいな話もしておく？

Author: たしかに。。。
`pgroonga_condition('keyword', ARRAY[weight1, weight2, ...], index_name => 'pgroonga_index')`を使うのは、`weight`を`0`にして検索対象から外す時に使う想定なので、`weight`を`0`にすると検索対象から外せることを書こうかと思います。

Author: `weight`を`0`にすると検索対象から外せることを書きました。

## Usage

Here are sample schema and data:

```sql
CREATE TABLE tags (
  name text PRIMARY KEY
);

CREATE INDEX pgroonga_tag_name_index ON tags
  USING pgroonga (name pgroonga_text_term_search_ops_v2)
  WITH (normalizers='NormalizerNFKC150("remove_symbol", true)');

INSERT INTO tags VALUES ('PostgreSQL');
INSERT INTO tags VALUES ('Groonga');
INSERT INTO tags VALUES ('PGroonga');
INSERT INTO tags VALUES ('pglogical');
```

インデックスサーチ時はインデックスに指定したオプションを使ってインデックスサーチ時の挙動をカスタマイズできます。
上のサンプルでは、`normalizers='...'`の部分でオプションを指定しています。

一方、インデックスサーチではなくシーケンシャルサーチが実行されると、インデックスに指定されているオプションは参照できません。
シーケンシャルサーチ時はどのインデックスを参照すればよいかという情報がないからです。

インデックスを参照できないということは、インデックスに設定されているノーマライザーやトークナイザーの情報を参照できず、適用もできません。
そのためシーケンシャルサーチ実行時は、インデックスサーチ実行時と検索結果が異なる可能性があります。
この問題を回避するためにシーケンシャルサーチ時に参照するインデックスを明示的に指定します。
`pgroonga_condition()`の`index_name => '...'`がそのための引数です。

次の例は、「-p_G」というキーワードで前方一致検索をしていますが、シーケンシャルサーチであっても、「PGroonga」と「pglogical」がヒットしています。

> このことから、シーケンシャルサーチ実行時でもインデックスに設定されている`NormalizerNFKC150("remove_symbol", true)`が参照できていることがわかります。

Reviewer 3: `remove_symbol` に関して、リンクを貼ってあげると良さそうかなと思いました！
- https://groonga.org/docs/reference/normalizers/normalizer_nfkc150.html#remove-symbol

過剰かもしれないですが、「インデックスが利用され、「-p_G」=> 「pG」にノーマライズされ検索されていることがわかる」みたいな補足を入れてあげるとわかりやすいかなと思いました。（リンクだけでも十分かもです。）

Author: リンクを貼りました。
また、`remove_symbol`の効果についても追記しました。

titleカラムに「Groonga」または「PostgreSQL」があるレコードの方がcontentカラムに「Groonga」または「PostgreSQL」がある方よりスコアーが高いことを確認できます。

`pgroonga_condition('keyword', ARRAY[weight1, weight2, ...], index_name => 'pgroonga_index')`は、検索対象のカラムを選択する場合に使います。
`weight`を`0`にすることで、対応するカラムを無視できます。次の例では、`content`カラムを無視して検索します。

`content`カラムも検索対象としている場合は、「-p_G」というキーワードで前方一致検索しているので、`'PGroonga', 'PostgreSQLの拡張機能です。'`がヒットするはずですが、次の例では、`content`カラムを無視して検索しているため、このレコードはヒットしていません。
次の例のように、検索対象のカラムを選択しつつ、インデックスに設定されているノーマライザーやトークナイザーを参照して検索したい場合には、この書き方を使ってください。

```sql
DROP TABLE IF EXISTS memos;
CREATE TABLE memos (
  title text,
  content text
);

CREATE INDEX pgroonga_memos_index
    ON memos
 USING pgroonga ((ARRAY[title, content]) pgroonga_text_array_term_search_ops_v2)
 WITH (normalizers='NormalizerNFKC150("remove_symbol", true)');

INSERT INTO memos VALUES ('PostgreSQL', 'PostgreSQLはリレーショナル・データベース管理システムです。');
INSERT INTO memos VALUES ('Groonga', 'Groongaは日本語対応の高速な全文検索エンジンです。');
INSERT INTO memos VALUES ('PGroonga', 'PostgreSQLの拡張機能です。');
INSERT INTO memos VALUES ('pglogical', 'pglogicalは、論理レプリケーションを実装しています。');

EXPLAIN ANALYZE VERBOSE SELECT *
  FROM memos
 WHERE ARRAY[title, content] &^ pgroonga_condition('-p_O',

>                                                    ARRAY[1, 1],

Reviewer: `EXPLAIN` では ` ARRAY[1, 1],` で、下部の実際の `SELECT` は `ARRAY[1, 0],` ですが、狙い通りですか？

Author: 間違いですね。。。
`EXPLAIN`のほうを`ARRAY[1, 0]`に変更しました。

## Usage

Here are sample schema and data:

```sql
CREATE TABLE tags (
  name text PRIMARY KEY
);

CREATE INDEX pgroonga_tag_name_index ON tags
  USING pgroonga (name pgroonga_text_term_search_ops_v2)
  WITH (normalizers='NormalizerNFKC150("remove_symbol", true)');

INSERT INTO tags VALUES ('PostgreSQL');
INSERT INTO tags VALUES ('Groonga');
INSERT INTO tags VALUES ('PGroonga');
INSERT INTO tags VALUES ('pglogical');
```

インデックスサーチ時はインデックスに指定したオプションを使ってインデックスサーチ時の挙動をカスタマイズできます。
上のサンプルでは、`normalizers='...'`の部分でオプションを指定しています。

一方、インデックスサーチではなくシーケンシャルサーチが実行されると、インデックスに指定されているオプションは参照できません。
シーケンシャルサーチ時はどのインデックスを参照すればよいかという情報がないからです。

> インデックスを参照できないということは、インデックスに設定されているノーマライザーやトークナイザーの情報を参照できず、適用もできません。

Reviewer 2: これ、いらなくない？

前の段落と同じことを言っていそう。

Author: 消しました。

そのためシーケンシャルサーチ実行時は、インデックスサーチ実行時と検索結果が異なる可能性があります。
この問題を回避するためにシーケンシャルサーチ時に参照するインデックスを明示的に指定します。
`pgroonga_condition()`の`index_name => '...'`がそのための引数です。

> 次の例は、「-p_G」というキーワードで前方一致検索をしており、インデックスには`NormalizerNFKC150("remove_symbol", true)`が設定されています。

Reviewer 2: ```suggestion
次の例は、「`-p_G`」というキーワードで前方一致検索をしており、インデックスには`NormalizerNFKC150("remove_symbol", true)`が設定されています。
```

> 次の例は、「-p_G」というキーワードで前方一致検索をしており、インデックスには`NormalizerNFKC150("remove_symbol", true)`が設定されています。

Reviewer 2: `-`から始まるクエリーはNOTっぽく見えるので使わない方がいいと思う。

Author: `-`ではなく`_`を使うようにしました。

> [`remove_symbol`][remove-symbol]は記号を無視するオプションなので、「-p_G」は「pg」にノーマライズされます。

Reviewer 2: ```suggestion
[`remove_symbol`][remove-symbol]は記号を無視するオプションなので、「`-p_G`」は「`pg`」にノーマライズされます。
```

> （大文字が小文字になっているのは、`NormalizerNFKC150`の挙動によるものです。）

Reviewer 2: 記号が削除されているのも`NormalizerNFKC150`の挙動っちゃ挙動だけどね。。。

Author: `remove_symbol`オプションの挙動ではないことを伝えたかったので、次の表現にしました。

```
（大文字が小文字になっているのは、`remove_symbol`オプションの挙動ではなく、`NormalizerNFKC150`のデフォルトの挙動によるものです。）
```

> そのため、このオプションが有効であれば、「PGroonga」と「pglogical」がヒットします。

Reviewer 2: ```suggestion
そのため、このオプションが有効であれば、「`PGroonga`」と「`pglogical`」がヒットします。
```

他にもあるけど面倒だからこれ以上はsuggestionしません。

Author: 一通り追加しました。

   Filter: (name &^ '(-p_G,,,,pgroonga_tag_name_index,)'::pgroonga_condition)
   Rows Removed by Filter: 2
 Planning Time: 0.871 ms
 Execution Time: 2.352 ms
(5 rows)

SELECT *
  FROM tags
 WHERE name &^ pgroonga_condition('-p_G',
                                  index_name => 'pgroonga_tag_name_index');
   name
-----------
 PGroonga
 pglogical
(2 rows)
```

`index_name`を指定しない場合（つまり、`NormalizerNFKC150("remove_symbol", true)`が参照できない場合）は、次のように「PGroonga」と「pglogical」はヒットしません。

```sql
EXPLAIN ANALYZE
SELECT *
  FROM tags
 WHERE name &^ pgroonga_condition('-p_G');
                                            QUERY PLAN
--------------------------------------------------------------------------------------------------
 Seq Scan on tags  (cost=0.00..1043.60 rows=1 width=32) (actual time=0.032..0.032 rows=0 loops=1)
   Filter: (name &^ '(-p_G,,,,,)'::pgroonga_condition)
   Rows Removed by Filter: 4
 Planning Time: 0.910 ms
 Execution Time: 0.053 ms
(5 rows)

SELECT *
  FROM tags
 WHERE name &^ pgroonga_condition('-p_G');

 name
------
(0 rows)
```

このように、`index_name`を指定することで、シーケンシャルサーチ実行時でもインデックスサーチ実行時でも検索結果が変わらないようにできます。

> また、カラム毎の`weight`も設定できます。

Reviewer 3: [nits]
Syntax で紹介している例の詳細を Usage で紹介しているので、
対応する詳細ごとにセクションをくぎってあげると見やすくなるかなと思いました。
```sql
pgroonga_condition('keyword', index_name => 'pgroonga_index')
pgroonga_condition('keyword', ARRAY[weight1, weight2, ...])
pgroonga_condition('keyword', ARRAY[weight1, weight2, ...], index_name => 'pgroonga_index')
```

下記のようなイメージをしています。
```md
## Using `index_name`

## Using `weight[]`

## Using `index_name` and `weight[]`
```

Author: 他の箇所がそういった書き方をしていないので、このドキュメントも他と合わせてセクションを区切るような書き方をしないほうが良いかなと思います。

Reviewer 3: 了解しました！

Reviewer 2: うーん、長いから https://pgroonga.github.io/reference/functions/pgroonga-query-expand.html みたいに分けた方がいいんじゃないかなぁ。

ただ、セクションレベルは`## Usage`より下げたほうがいいと思う。

```markdown
## Usage
### `index_name`
### `weights`
### `index_name` and `weights`
```

   Filter: (name &^ '(_p_G,,,,pgroonga_tag_name_index,)'::pgroonga_condition)
   Rows Removed by Filter: 2
 Planning Time: 0.871 ms
 Execution Time: 2.352 ms
(5 rows)

SELECT *
  FROM tags
 WHERE name &^ pgroonga_condition('_p_G',
                                  index_name => 'pgroonga_tag_name_index');
   name
-----------
 PGroonga
 pglogical
(2 rows)
```

`index_name`を指定しない場合（つまり、`NormalizerNFKC150("remove_symbol", true)`が参照できない場合）は、次のように「`PGroonga`」と「`pglogical`」はヒットしません。

```sql
EXPLAIN ANALYZE
SELECT *
  FROM tags
 WHERE name &^ pgroonga_condition('_p_G');
                                            QUERY PLAN
--------------------------------------------------------------------------------------------------
 Seq Scan on tags  (cost=0.00..1043.60 rows=1 width=32) (actual time=0.032..0.032 rows=0 loops=1)
   Filter: (name &^ '(_p_G,,,,,)'::pgroonga_condition)
   Rows Removed by Filter: 4
 Planning Time: 0.910 ms
 Execution Time: 0.053 ms
(5 rows)

SELECT *
  FROM tags
 WHERE name &^ pgroonga_condition('_p_G');

 name
------
(0 rows)
```

このように、`index_name`を指定することで、シーケンシャルサーチ実行時でもインデックスサーチ実行時でも検索結果が変わらないようにできます。

> ### Specify `weight`

Reviewer 3: ```suggestion
### Specify `weights`
```

Reviewer 2: これ、適用して。

> また、カラム毎の`weight`も設定できます。
> つまり、「タイトルは本文よりも重要」を実現できます。

Reviewer 3: セクションで区切った場合に、新しく文章をはじめる形にすると違和感がないかなとおもいました。
あまり良い冒頭をイメージできませんでしたが、下記のような感じを想定していますmm
```suggestion
カラム毎に異なる重要度（weights）を設定する方法を紹介します。
これにより、特定のカラムをより重要視することが可能になり、より柔軟な検索結果の調整が行えます。
例えば、「タイトルは本文よりも重要視する」などを実現できます。
```

Reviewer 3: [nits]
各サブセクションの構成を同じだと統一感があって読みやすいかなと思いました。

```
### Section title

- 使い方の説明
- Here are sample schema and data:
- 具体例

```

Author: セクションごとに、新しく文章を始める形にしてみました。

Author: 各セクションを下の構成にしてみました。

```
### Section title

- 使い方の説明
- Here are sample schema and data:
- 具体例
```

## Usage

### Specify `index_name`

> シーケンシャルサーチ実行時でも、インデックスに指定したノーマライザーやトークナイザーのオプションを使って検索する方法を紹介します。

Reviewer 4: Introducing how to search with normalizer and tokeniser options indicated in the index  while sequential search is executed.

`pgroonga_condition('keyword', index_name => 'pgroonga_index')`を使います。
`index_name`にノーマライザーやトークナイザーを指定したインデックスの名前を指定します。

Here are sample schema and data:

```sql
CREATE TABLE tags (
  name text PRIMARY KEY
);

CREATE INDEX pgroonga_tag_name_index ON tags
  USING pgroonga (name pgroonga_text_term_search_ops_v2)
  WITH (normalizers='NormalizerNFKC150("remove_symbol", true)');

INSERT INTO tags VALUES ('PostgreSQL');
INSERT INTO tags VALUES ('Groonga');
INSERT INTO tags VALUES ('PGroonga');
INSERT INTO tags VALUES ('pglogical');
```

インデックスサーチ時はインデックスに指定したオプションを使ってインデックスサーチ時の挙動をカスタマイズできます。
上のサンプルでは、`normalizers='...'`の部分でオプションを指定しています。

一方、インデックスサーチではなくシーケンシャルサーチが実行されると、インデックスに指定されているオプションは参照できません。
シーケンシャルサーチ時はどのインデックスを参照すればよいかという情報がないからです。

> そのためシーケンシャルサーチ実行時は、インデックスサーチ実行時と検索結果が異なる可能性があります。
> この問題を回避するためにシーケンシャルサーチ時に参照するインデックスを明示的に指定します。
> `pgroonga_condition()`の`index_name => '...'`がそのための引数です。
> 

Reviewer 3: ```suggestion
Because of that, there is a possibility that the search results may differ between a sequential search and an index search.
In order to avoid this issue, explicitly specify which index should be referenced at a sequential search.
The `index_name => '...'` argument in `pgroonga_condition()` is used for that.
```

> 次の例は、「`_p_G`」というキーワードで前方一致検索をしており、インデックスには`NormalizerNFKC150("remove_symbol", true)`が設定されています。
> [`remove_symbol`][remove-symbol]は記号を無視するオプションなので、「`_p_G`」は「`pg`」にノーマライズされます。
> （大文字が小文字になっているのは、`remove_symbol`オプションの挙動ではなく、`NormalizerNFKC150`のデフォルトの挙動によるものです。）
> そのため、このオプションが有効であれば、「`PGroonga`」と「`pglogical`」がヒットします。
> 
> 次の例は、シーケンシャルサーチですが、「`PGroonga`」と「`pglogical`」がヒットしていることが確認できます。
> このことから、シーケンシャルサーチ実行時でもインデックスに設定されている`NormalizerNFKC150("remove_symbol", true)`が参照できていることが確認できます。

Reviewer 4: ```suggestion

Next example execute a prefix search with the keyword "`_p_G`" while `NormalizerNFKC150("remove_symbol", true)` is configured to the index.   
Since [`remove_symbol`][remove-symbol] is the option to ignore the symbols, "`_p_G`" would be normalized to "`pg`".
( The reason why a capital letter `G` become a lower case latter `g` is not a behavior `remove_symbol` option, and is a default behavior of  `NormalizerNFKC150`.)
Therefore, "`PGroonga`" and "`pglogical`" should be hit while this option is valid. 

You can see "`PGroonga`" and "`pglogical`" are hit as result for the sequential search. 
This result shows that `NormalizerNFKC150("remove_symbol" configured to the index is referenced while the sequential search is executed.  

```

検索をしているというのを検索を実行しているという感じで`execute`にしていますが、「プログラムしている」みたいな感じのほうがいいのでしょうか・・・設定している？という感じでしょうか・・・。

Author: 検索をしているというのを検索を実行しているという感じで大丈夫です。

Reviewer 3: ```suggestion
The next example executes a prefix search with the keyword "`_p_G`` while `NormalizerNFKC150("remove_symbol", true)` is specified in the index.
Since [`remove_symbol`][remove-symbol] is an option to ignore symbols, "`_p_G`" is normalized to "`pg`".
(The reason why the capital letter "`G`" becomes a lowercase "`g`" is not due to the `remove_symbol` option, but rather the default behavior of `NormalizerNFKC150`.)
Therefore, both "`PGroonga`" and "`pglogical`" should be hit while this option is active.
You can see that "`PGroonga`" and "`pglogical`" are hit as a result of the sequential search.
This result shows that `NormalizerNFKC150("remove_symbol" specified in the index is referenced even during sequential search execution.
```

```sql
EXPLAIN ANALYZE
SELECT *
  FROM tags
 WHERE name &^ pgroonga_condition('_p_G',
                                  index_name => 'pgroonga_tag_name_index');
                                            QUERY PLAN
--------------------------------------------------------------------------------------------------
 Seq Scan on tags  (cost=0.00..1043.60 rows=1 width=32) (actual time=2.267..2.336 rows=2 loops=1)
   Filter: (name &^ '(_p_G,,,,pgroonga_tag_name_index,)'::pgroonga_condition)
   Rows Removed by Filter: 2
 Planning Time: 0.871 ms
 Execution Time: 2.352 ms
(5 rows)

SELECT *
  FROM tags
 WHERE name &^ pgroonga_condition('_p_G',
                                  index_name => 'pgroonga_tag_name_index');
   name
-----------
 PGroonga
 pglogical
(2 rows)
```

> `index_name`を指定しない場合（つまり、`NormalizerNFKC150("remove_symbol", true)`が参照できない場合）は、次のように「`PGroonga`」と「`pglogical`」はヒットしません。

Reviewer 3: ```suggestion
As you can see in next, "`PGroonga`"and "`pglogical`" would not be hit when `index_name` is not specified. (This means that `NormalizerNFKC150("remove_symbol", true)` cannot be referenced. )
```

```sql
EXPLAIN ANALYZE
SELECT *
  FROM tags
 WHERE name &^ pgroonga_condition('_p_G');
                                            QUERY PLAN
--------------------------------------------------------------------------------------------------
 Seq Scan on tags  (cost=0.00..1043.60 rows=1 width=32) (actual time=0.032..0.032 rows=0 loops=1)
   Filter: (name &^ '(_p_G,,,,,)'::pgroonga_condition)
   Rows Removed by Filter: 4
 Planning Time: 0.910 ms
 Execution Time: 0.053 ms
(5 rows)

SELECT *
  FROM tags
 WHERE name &^ pgroonga_condition('_p_G');

 name
------
(0 rows)
```

> このように、`index_name`を指定することで、シーケンシャルサーチ実行時でもインデックスサーチ実行時でも検索結果が変わらないようにできます。

Reviewer 3: ```suggestion
In this way, by specifying `index_name`, you can ensure that the search result remain the same whether executing a sequential search or an index search.
```

### Specify `weights`

> カラム毎に異なるweight（重要度）を設定する方法を紹介します。

Reviewer 3: ```suggestion
カラム毎に異なる`weight`（重要度）を設定する方法を紹介します。
```

> カラム毎に異なるweight（重要度）を設定する方法を紹介します。
> これにより、「タイトルを本文よりも重要視する」を実現できます。
> 
> `pgroonga_condition('keyword', ARRAY[weight1, weight2, ...])` を使います。
> `weight1`、 `weight2`でカラム毎の重要度を指定します。

Reviewer 4: ```suggestion

Introducing how to configure the different weight for each columns.
This let your title to be weighted more than the main text. 

Use  `pgroonga_condition('keyword', ARRAY[weight1, weight2, ...])`.
`weight1`, `weight2` configure the weight of each columns.
```

Reviewer 3: ```suggestion
Introducing how to specify different weights for each column.
This allows the title to be weighted more than the main text.

Use `pgroonga_condition('keyword', ARRAY[weight1, weight2, ...])`.
`weight1`, `weight2`, and so on specify the weights for each column.
```

Here are sample schema and data:

```sql
DROP TABLE IF EXISTS memos;
CREATE TABLE memos (
  title text,
  content text
);

CREATE INDEX pgroonga_memos_index
    ON memos
 USING pgroonga ((ARRAY[title, content]));

> 
> INSERT INTO memos VALUES ('PostgreSQL', 'PostgreSQLはリレーショナル・データベース管理システムです。');
> INSERT INTO memos VALUES ('Groonga', 'Groongaは日本語対応の高速な全文検索エンジンです。');
> INSERT INTO memos VALUES ('PGroonga', 'PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。');
> INSERT INTO memos VALUES ('コマンドライン', 'groongaコマンドがあります。');
> ```

Reviewer 4: ```suggestion

INSERT INTO memos VALUES ('PostgreSQL', 'PostgreSQL is a relational database management system.');
INSERT INTO memos VALUES ('Groonga', 'Groongais the fast full text search engine accommodated for Japanese.');
INSERT INTO memos VALUES ('PGroonga', 'PGroonga is an add-on of the PostgreSQL to use Groonga as the index.');
INSERT INTO memos VALUES ('command line', ' There is a groonga command.');
```
```

Reviewer 3: ```suggestion
INSERT INTO memos VALUES ('PostgreSQL', 'PostgreSQL is a relational database management system.');
INSERT INTO memos VALUES ('Groonga', 'Groonga is the fast full text search engine optimized  for Japanese.');
INSERT INTO memos VALUES ('PGroonga', 'PGroonga is an extension for PostgreSQL to use Groonga as the index.');
INSERT INTO memos VALUES ('command line', 'There is a groonga command.');
```

> 指定したクエリーによりマッチしたレコードを探すためには[`pgroonga_score function`][pgroonga-score-function]を使えます。

Reviewer 4: ```suggestion
[`pgroonga_score function`][pgroonga-score-function] can be used to search a record matched by specified query.
```

Reviewer 3: ```suggestion
[`pgroonga_score function`][pgroonga-score-function] can be used to search for records that match by the specified query.
```

```sql
SELECT *, pgroonga_score(tableoid, ctid) AS score
  FROM memos
 WHERE ARRAY[title, content] &@~
       pgroonga_condition('Groonga OR PostgreSQL', ARRAY[5, 1])
 ORDER BY score DESC;
      title      |                                  content                                  | score 
----------------+---------------------------------------------------------------------------+-------

>  Groonga        | Groongaは日本語対応の高速な全文検索エンジンです。                         |     6
>  PostgreSQL     | PostgreSQLはリレーショナル・データベース管理システムです。                |     6
>  PGroonga       | PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。 |     2
>  コマンドライン | groongaコマンドがあります。                                               |     1

Reviewer 4: ```suggestion
 Groonga         | Groonga is a fast full text search engine accommodated for Japanese.                         |     6
 PostgreSQL     | PostgreSQL is a relational database management system.                |     6
 PGroonga       | PGroonga is an add-on of the PostgreSQL to use Groonga as the index. |     2
 command line |  There is a groonga command.                                          |     1
```

Reviewer 3: ```suggestion
 Groonga        | Groonga is a fast full text search engine optimized for Japanese.                         |     6
 PostgreSQL     | PostgreSQL is a relational database management system.                |     6
 PGroonga       | PGroonga is an extension for PostgreSQL to use Groonga as the index. |     2
command line | There is a groonga command. | 1
```

(4 rows)
```

> 上の例では、`ARRAY[title, content] &@~ pgroonga_condition('Groonga OR PostgreSQL', ARRAY[5, 1])`と指定しているので、タイトルが本文より5倍重要としています。
> `title`カラムに「`Groonga`」または「`PostgreSQL`」があるレコードの方が`content`カラムに「`Groonga`」または「`PostgreSQL`」がある方よりスコアーが高いことを確認できます。

Reviewer 4: I
```suggestion
In above example, the title is 5 times weighted that the main text because `ARRAY[title, content] &@~ pgroonga_condition('Groonga OR PostgreSQL', ARRAY[5, 1])` is specified.
You can see the score is higher for the records with`Groonga` or `PostgreSQL` in the `title` column compare to the records with `Groonga` or `PostgreSQL`  in the `content` column.
```

Reviewer 3: ```suggestion
In above example, the title is 5 times weighted that the main text because `ARRAY[title, content] &@~ pgroonga_condition('Groonga OR PostgreSQL', ARRAY[5, 1])` is specified.
You can see the score is higher for the records with`Groonga` or `PostgreSQL` in the `title` column compared to records with `Groonga` or `PostgreSQL` in the `content` column.
```

### Exclude from search target

> 特定のカラムを検索対象から除外して検索する方法を紹介します。
> 
> `pgroonga_condition('keyword', ARRAY[weight1, weight2, ...])`を使います。
> 検索対象から除外したいカラムに対応する`weight`に`0`を指定します。

Reviewer 4: ```suggestion
Introducing how to search without specific columns as the search target.

Use `pgroonga_condition('keyword', ARRAY[weight1, weight2, ...])`.
You need to specify `0` to `weight` of the column that would be excluded from the search target.
```

Here are sample schema and data:

```sql
DROP TABLE IF EXISTS memos;
CREATE TABLE memos (
  title text,
  content text
);

CREATE INDEX pgroonga_memos_index
    ON memos
 USING pgroonga ((ARRAY[title, content]));

> INSERT INTO memos VALUES ('PostgreSQL', 'PostgreSQLはリレーショナル・データベース管理システムです。');
> INSERT INTO memos VALUES ('Groonga', 'Groongaは日本語対応の高速な全文検索エンジンです。');
> INSERT INTO memos VALUES ('PGroonga', 'PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。');
> INSERT INTO memos VALUES ('コマンドライン', 'groongaコマンドがあります。');

Reviewer 4: ```suggestion
INSERT INTO memos VALUES ('PostgreSQL', 'PostgreSQL is a relational database management system.');
INSERT INTO memos VALUES ('Groonga', 'Groonga is a fast full text search engine accommodated for Japanese');
INSERT INTO memos VALUES ('PGroonga', 'PGroonga', 'PGroonga is an add-on of the PostgreSQL to use Groonga as the index.');
INSERT INTO memos VALUES ('command line', ' There is a groonga command.');
```

Reviewer 3: ```suggestion
INSERT INTO memos VALUES ('PostgreSQL', 'PostgreSQL is a relational database management system.');
INSERT INTO memos VALUES ('Groonga', 'Groonga is a fast full text search engine optimized for Japanese');
INSERT INTO memos VALUES ('PGroonga', 'PGroonga', 'PGroonga is an extension for PostgreSQL to use Groonga as the index.');
INSERT INTO memos VALUES ('command line', ' There is a groonga command.');
```

```

> 次の例では、`content`カラムを検索対象から除外しています。
> 「`拡張`」というキーワードで全文検索しているので、`content`カラムを検索対象としていれば、`'PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。'`がヒットするはずですが、このレコードはヒットしていません。
> このことから、`content`カラムが検索対象から除外されていることを確認できます。

Reviewer 4: ```suggestion
In the next sample, `content` column is excluded from the search target. 
The record, 'PGroonga is an add-on of the PostgreSQL to use Groonga as the index.', should be hit if the `content` column is the search target with the search keyword is `add-on`, but the record is not hit.
You can see that the `content` column is not set as the search target.
```

Reviewer 3: ```suggestion
In the next example, the `content` column is excluded from the search target. 
The record, `'PGroonga is an extension for PostgreSQL to use Groonga as the index.'`, should be hit if the `content` column is included in the search target with the search keyword is "`extension`", but the record is not hit.
You can see that the `content` column is excluded from the search target.
```

```sql
SELECT *
  FROM memos
 WHERE ARRAY[title, content] &@~
       pgroonga_condition('拡張', ARRAY[1, 0]);
 title | content
-------+---------
(0 rows)
```

> 次のように、`content`カラムを検索対象にすると`'PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。'`がヒットします。

Reviewer 4: ```suggestion
As `content` column is set as the search target in next, the record, PGroonga is an add-on of the PostgreSQL to use Groonga as the index.', is hit.
```

Reviewer 3: ```suggestion
As the `content` column is set as the search target in the next example, the record `'PGroonga is an extension for PostgreSQL to use Groonga as an index.'` is hit.
```

```sql
SELECT *
  FROM memos
 WHERE ARRAY[title, content] &@~
       pgroonga_condition('拡張', ARRAY[1, 1]);
  title   |                                  content
----------+---------------------------------------------------------------------------

>  PGroonga | PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。

Reviewer 4: ```suggestion
 PGroonga | PGroonga is an add-on of the PostgreSQL to use Groonga as the index.
```

Reviewer 3: ```suggestion
 PGroonga | PGroonga is an extension for PostgreSQL to use Groonga as the index.
```

 Seq Scan on tags  (cost=0.00..1043.60 rows=1 width=32) (actual time=0.032..0.032 rows=0 loops=1)
   Filter: (name &^ '(_p_G,,,,,)'::pgroonga_condition)
   Rows Removed by Filter: 4
 Planning Time: 0.910 ms
 Execution Time: 0.053 ms
(5 rows)

SELECT *
  FROM tags
 WHERE name &^ pgroonga_condition('_p_G');

 name
------
(0 rows)
```

このように、`index_name`を指定することで、シーケンシャルサーチ実行時でもインデックスサーチ実行時でも検索結果が変わらないようにできます。

### Specify `weight`

カラム毎に異なるweight（重要度）を設定する方法を紹介します。
これにより、「タイトルを本文よりも重要視する」を実現できます。

`pgroonga_condition('keyword', ARRAY[weight1, weight2, ...])` を使います。
`weight1`、 `weight2`でカラム毎の重要度を指定します。

Here are sample schema and data:

```sql
DROP TABLE IF EXISTS memos;
CREATE TABLE memos (
  title text,
  content text
);

CREATE INDEX pgroonga_memos_index
    ON memos
 USING pgroonga ((ARRAY[title, content]));
```

```sql
INSERT INTO memos VALUES ('PostgreSQL', 'PostgreSQLはリレーショナル・データベース管理システムです。');
INSERT INTO memos VALUES ('Groonga', 'Groongaは日本語対応の高速な全文検索エンジンです。');
INSERT INTO memos VALUES ('PGroonga', 'PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。');
INSERT INTO memos VALUES ('コマンドライン', 'groongaコマンドがあります。');
```

> より指定したクエリーにマッチしたレコードを探すためには[`pgroonga_score function`][pgroonga-score-function]を使えます。

Reviewer 2: ```suggestion
指定したクエリーによりマッチしたレコードを探すためには[`pgroonga_score function`][pgroonga-score-function]を使えます。
```

```sql
SELECT *, pgroonga_score(tableoid, ctid) AS score
  FROM memos
 WHERE ARRAY[title, content] &@~
       pgroonga_condition('Groonga OR PostgreSQL', ARRAY[5, 1])
 ORDER BY score DESC;
      title      |                                  content                                  | score 
----------------+---------------------------------------------------------------------------+-------
 Groonga        | Groongaは日本語対応の高速な全文検索エンジンです。                         |     6
 PostgreSQL     | PostgreSQLはリレーショナル・データベース管理システムです。                |     6
 PGroonga       | PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。 |     2
 コマンドライン | groongaコマンドがあります。                                               |     1
(4 rows)
```

上の例では、`ARRAY[title, content] &@~ pgroonga_condition('Groonga OR PostgreSQL', ARRAY[5, 1])`と指定しているので、タイトルが本文より5倍重要としています。
`title`カラムに「`Groonga`」または「`PostgreSQL`」があるレコードの方が`content`カラムに「`Groonga`」または「`PostgreSQL`」がある方よりスコアーが高いことを確認できます。

> ### Specify `index_name` and `weight`

Reviewer 2: ```suggestion
### Specify `index_name` and `weights`
```

> 特定のカラムを検索対象から除外しつつ、シーケンシャルサーチ実行時でも、インデックスに指定したノーマライザーやトークナイザーのオプションを使って検索する方法を紹介します。

Reviewer 2: 日本語が複雑で英語を作るときに苦労しそう。

Author: 除外だけで1セクションにしたので、「シーケンシャルサーチ実行時でも...」の部分を削除して、「特定のカラムを検索対象から除外して検索する方法を紹介します。」にしました。

`pgroonga_condition('keyword', ARRAY[weight1, weight2, ...], index_name => 'pgroonga_index')`を使います。

> 検索対象から除外したいカラムに対応する`weight`に`0`を指定します。`index_name`には、ノーマライザーやトークナイザーを指定したインデックスの名前を指定します。

Reviewer 2: うーん、除外とインデックス指定は独立だから、除外だけで1セクションにしたら？

一緒に説明したい？

Author: 一緒に説明したくないので、除外だけで1セクションにしてみました。

```sql
SELECT *, pgroonga_score(tableoid, ctid) AS score
  FROM memos
 WHERE ARRAY[title, content] &@~
       pgroonga_condition('Groonga OR PostgreSQL', ARRAY[5, 1])
 ORDER BY score DESC;
      title      |                                  content                                  | score 
----------------+---------------------------------------------------------------------------+-------
 Groonga        | Groongaは日本語対応の高速な全文検索エンジンです。                         |     6
 PostgreSQL     | PostgreSQLはリレーショナル・データベース管理システムです。                |     6
 PGroonga       | PGroongaはインデックスとしてGroongaを使うためのPostgreSQLの拡張機能です。 |     2
 コマンドライン | groongaコマンドがあります。                                               |     1
(4 rows)
```

上の例では、`ARRAY[title, content] &@~ pgroonga_condition('Groonga OR PostgreSQL', ARRAY[5, 1])`と指定しているので、タイトルが本文より5倍重要としています。
`title`カラムに「`Groonga`」または「`PostgreSQL`」があるレコードの方が`content`カラムに「`Groonga`」または「`PostgreSQL`」がある方よりスコアーが高いことを確認できます。

### Specify `index_name` and `weight`

検索対象のカラムを選択しつつ、シーケンシャルサーチ実行時でも、インデックスに指定したノーマライザーやトークナイザーのオプションを使って検索する方法を紹介します。

`pgroonga_condition('keyword', ARRAY[weight1, weight2, ...], index_name => 'pgroonga_index')`を使います。

> 無視するカラムに対応する`weight`に`0`を指定します。`index_name`にノーマライザーやトークナイザーを指定したインデックスの名前を指定します。

Reviewer 3: 「検索対象のカラムを選択」から「特定のカラムを無視する」の流れの間にクッションがあると、
何故いきなり「カラムを無視する」話になったのかが、分かって良さそうに思います。
いい感じの文章を思いつけなかったのですが、下記のようなイメージでいます。

```suggestion
`weight` を操作し特定のカラムを重視したり無視したりすることで検索対象のカラムの選択が可能です。
更に合わせて、シーケンシャルサーチ時に利用するインデックスを明示的に指定可能です。
```

Author: シーケンシャルサーチのときは「特定のカラムを重視する」はできないので、検索対象のカラムにするか/しないかだけを制御できます。

A,B,Cというカラムがあって、Bだけ検索対象にしない、A,Cを検索対象にする、という操作を「カラムの選択」という意味で使いました。
「特定のカラムを無視する」ではなくて、「特定のカラムを検索対象にしない」という表現に変えると良いですか？

Reviewer 3: > シーケンシャルサーチのときは「特定のカラムを重視する」はできないので、検索対象のカラムにするか/しないかだけを制御できます。

あぁなるほどです！上記を分かっていなかったです。
いま説明していただいた内容を入れるとすごくわかりやすくなる気がします。

Author: 「シーケンシャルサーチのときは「特定のカラムを重視する」はできない」ことを説明し始めると、話題をまとめきれないので、「特定のカラムを除外したい」+「シーケンシャルサーチの時にインデックスの情報を参照したい」場合に使う、ということだけを説明する内容にしました。どうでしょうか？

Reviewer 3: 良さそうに思いました。ありがとうございます！

Here are sample schema and data:

```sql
DROP TABLE IF EXISTS memos;
CREATE TABLE memos (
  title text,
  content text
);

CREATE INDEX pgroonga_memos_index
    ON memos
 USING pgroonga ((ARRAY[title, content]) pgroonga_text_array_term_search_ops_v2)
 WITH (normalizers='NormalizerNFKC150("remove_symbol", true)');

INSERT INTO memos VALUES ('PostgreSQL', 'PostgreSQLはリレーショナル・データベース管理システムです。');
INSERT INTO memos VALUES ('Groonga', 'Groongaは日本語対応の高速な全文検索エンジンです。');
INSERT INTO memos VALUES ('PGroonga', 'PostgreSQLの拡張機能です。');
INSERT INTO memos VALUES ('pglogical', 'pglogicalは、論理レプリケーションを実装しています。');
```

次の例では、`content`カラムを検索対象から除外しています。

> 次の例では、「`_p_O`」というキーワードで前方一致検索しているので、`content`カラムを検索対象としていれば、`'PGroonga', 'PostgreSQLの拡張機能です。'`がヒットするはずですが、このレコードはヒットしていません。このことから、`content`カラムが検索対象から除外されていることを確認できます。

Reviewer 3: `次の例では`が２度並んでいると少し冗長かなと思ったので、一文にしてしまうのはいかがでしょうか？

Author: 一文にしました。

## Usage

Here are sample schema and data:

```sql
CREATE TABLE memos (
  id integer,
  title text,
  content text
);

CREATE INDEX pgroonga_memos_index
          ON memos
       USING pgroonga (title)
        WITH (normalizers='NormalizerNFKC150("unify_katakana_v_sounds", true)');

INSERT INTO memos VALUES (1, 'ヴァイオリン', E'Let\'s play violin!');
```

インデックスサーチ時はインデックスに指定したオプションで検索結果をカスタマイズできます。
上のサンプルでは、`normalizers='...'`の部分でオプションを指定しています。
一方、インデックスサーチではなくシーケンシャルサーチが実行されると、PGroongaのインデックスに指定されているオプションを参照できません。

> シーケンシャルサーチ時はどのインデックスを参照すればよいかという情報がないからです。

Reviewer 3: インデックス利用時とシーケンシャルサーチでインデックスを利用できない場合の例の対比があると、
わかりやすいかなと思いました 🙏🏾 
```sql
# インデックが利用できて参照できる例
SELECT *
  FROM memos
 WHERE title &@~ pgroonga_condition('バイオリン',
                                    index_name => 'pgroonga_memos_index');

SET enable_seqscan = on;
SET enable_indexscan = off;
SET enable_bitmapscan = off;

# シーケンシャルサーチになり参照できない例
SELECT *
  FROM memos
 WHERE title &@~ pgroonga_condition('バイオリン',
                                    index_name => 'pgroonga_memos_index');
```

Author: `index_name`を指定しない場合で、シーケンシャルサーチの時の例を追加してみました。
`index_name`を指定するとインデックスの情報を参照できる、としているので、`index_name`を指定しない場合を対比することにしました。

インデックスを参照できないということは、インデックスに設定されているノーマライザーやトークナイザーの情報も参照できません。

> シーケンシャルサーチ実行時は、インデックスに設定されたノーマライザーとトークナイザーを適用できないため、インデックスサーチ実行時と検索結果が異なる可能性があります。

Reviewer: 109行目と110行目の内容が重複している印象があります。
（「ノーマライザーやトークナイザーの情報も参照・適用できません」という内容に重複感があります。）

例えば以下のような感じにするとスッキリしそうです。

```
インデックスを参照できないということは、インデックスに設定されているノーマライザーやトークナイザーの情報を参照できず、適用もできません。
そのためシーケンシャルサーチ実行時は、インデックスサーチ実行時と検索結果が異なる可能性があります。
```

Author: 提案してもらった文に書き換えました。
